enhancements/aws-pod-identity.md

```diff
@@ -0,0 +1,569 @@
+manifest render time
+
+### Implementation Details/Notes/Constraints [optional]
+
+#### OIDC Provider Creation and Reconciliation (Phase 1)
+The name of the bucket and OIDC provider can be generated from the
+`status.infrastructureName` in the `Infrastructure` global config.  The prefix
+`sjenning-abcde` will be used in this document.
+
+The following is a bash script that does what the OIDC controller will need to
+do to create the `discovery.json` and `keys.json`.  The controller will need to
+watch `bound-sa-token-signing-certs` for change and update `keys.json`
+```bash
+#!/bin/bash
+
+set -xe
+
+export S3_BUCKET=sjenning-abcde-oidc-provider
+export AWS_REGION=us-west-1
+
+# Extract the serviceaccount keypair from cluster
+PKCS_KEY="sa-signer.pub"
+oc get -n openshift-kube-apiserver configmap -ojson bound-sa-token-signing-certs | jq -r '.data["service-account-001.pub"]' > $PKCS_KEY
+
+_bucket_name=$(aws s3api list-buckets  --query "Buckets[?Name=='$S3_BUCKET'].Name | [0]" --out text)
+if [ $_bucket_name == "None" ]; then
+    aws s3api create-bucket --bucket $S3_BUCKET --create-bucket-configuration LocationConstraint=$AWS_REGION
+fi
+echo "export S3_BUCKET=$S3_BUCKET"
+export HOSTNAME=s3-$AWS_REGION.amazonaws.com
+export ISSUER_HOSTPATH=$HOSTNAME/$S3_BUCKET
+
+cat <<EOF > discovery.json
+{
+    "issuer": "https://$ISSUER_HOSTPATH/",
+    "jwks_uri": "https://$ISSUER_HOSTPATH/keys.json",
+    "authorization_endpoint": "urn:kubernetes:programmatic_authorization",
+    "response_types_supported": [
+        "id_token"
+    ],
+    "subject_types_supported": [
+        "public"
+    ],
+    "id_token_signing_alg_values_supported": [
+        "RS256"
+    ],
+    "claims_supported": [
+        "sub",
+        "iss"
+    ]
+}
+EOF
+
+./self-hosted -key $PKCS_KEY  | jq '.keys += [.keys[0]] | .keys[1].kid = ""' > keys.json
+
+aws s3 cp --acl public-read ./discovery.json s3://$S3_BUCKET/.well-known/openid-configuration
+aws s3 cp --acl public-read ./keys.json s3://$S3_BUCKET/keys.json
+
+curl https://$ISSUER_HOSTPATH/.well-known/openid-configuration
+curl https://$ISSUER_HOSTPATH/keys.json
+```
+The mentioned `self-hosted` binary is from
+[here](https://github.com/openshift/aws-pod-identity-webhook/tree/master/hack/self-hosted)
+
+#### Modifying initial `Authentication` global config (Phase 1)
+
+In order for any of this to work, the `serviceAccountIssuer` must be set in the `Authentication` global config
+```yaml
+apiVersion: config.openshift.io/v1
+kind: Authentication
+metadata:
+  annotations:
+    release.openshift.io/create-only: "true"
+  name: cluster
+spec:
+  serviceAccountIssuer: https://s3-us-west-1.amazonaws.com/sjenning-abcde-oidc-provider
+status:
+  integratedOAuthMetadata:
+    name: oauth-openshift
+```
+Ideally, new 4.5 clusters deployed on platform `aws` would roll out with this set day-1.
+
+For day-2 deployments, changing this results in a redeployment of
```

Comment:
Will the first release of CCO (and all future ones) with support for pod-identiy be responsible for performing this modification of the Authentication object (so that all 4.5+ clusters are in pod-identity-capable-mode)?